x-pack/plugins/rule_registry/server/rule_data_client/index.ts

```diff
@@ -70,9 +70,10 @@ export class RuleDataClient implements IRuleDataClient {
     };
   }
 
-  getWriter(options: { namespace?: string } = {}): RuleDataWriter {
+  async getWriter(options: { namespace?: string } = {}): Promise<RuleDataWriter> {
     const { namespace } = options;
     const alias = getNamespacedAlias({ alias: this.options.alias, namespace });
+    await this.rolloverAliasIfNeeded({ namespace });
```

Comment:
Having `await this.rolloverAliasIfNeeded({ namespace });` inside `getWriter` could carry significant performance penalty. 

Let's consider solution-side code that performs many writes during a request or rule execution. A single log write operation could look like that:

```js
await ruleDataClientInstance.getWriter({ namespace }).bulk(logsToWrite);
```

And I can imagine a rule that performs tens of writes like that during the execution.  From the solution developer's perspective, nothing bad happens; they write logs N times. But under the hood, every write request is accompanied by at least three additional requests (total requests = N*4):

```js
await clusterClient.indices.rollover({ dry_run: true })
await clusterClient.indices.get()
await clusterClient.indices.simulateIndexTemplate()
```

To avoid making these extra requests on every write, the developer should instantiate the writer and reuse it. But I think it's (a) not always convenient to do, and (b) the developer should know implementation details of `getWriter` and be aware of the possible pitfall.

We can envisage use cases as above and protect API users from harming performance by having a map of instantiated writers (`Map<Namespace, RuleDataWriter>()`) and returning a writer for a given namespace before even calling `rolloverAliasIfNeeded` if it is already in the map. 

What do you think?

Reply:
> To avoid making these extra requests on every write, the developer should instantiate the writer and reuse it.

That's I'd say how it should be done. If you're in an executor function, you will get a `namespace` from the rule parameters and get the writer. Then you'll pass it down to your classes and functions as a dependency to write events.

Same if you need to log stuff from route handlers.

> having a map of instantiated writers (Map<Namespace, RuleDataWriter>())

Hehe, that's similar to what I had in my draft :) I think this could be handy in general if someone would always use the API like that: `await ruleDataClientInstance.getWriter({ namespace }).bulk(logsToWrite);`, but maybe YAGNI.

I'd say let's consider that in a follow-up PR.

Reply:
Yeah this is an area that is a little bit tricky. I had in mind adding some dev docs to explain the expected usage pattern (create a writer inside the alert executor and reuse it for the duration of the executor).

Typescript does do a little bit more to guide devs towards that pattern naturally - since `getWriter` is async, you have to do something like 
```js
const writer = await ruleDataClientInstance.getWriter({ namespace });
await writer.bulk(logsToWrite);
```
which might help clue devs in that writers are meant to be reused.

On the flip side, the potential issue with caching the writers and therefore only running the rollover logic once at instantiation time is that if the mappings change while Kibana is running then we won't detect it unless we add a way to manually re-check the mapping versions for an index. Our built-in mappings won't change this way, but we've had multiple cases of advanced users modifying the mappings to add their own custom fields. If we do add the user-defined templates (https://github.com/elastic/kibana/issues/101016#issuecomment-857593741) then those would be changing while Kibana is running - so we'd want to check the template versions frequently to make sure changes to those templates get propagated to the indices.

So in my mind the RuleDataWriter is designed to live for a few seconds, e.g. for the duration of a rule execution, after which it should be discarded and a new one instantiated. My hope was that with docs we can provide simple guidance about creating and using a single writer within the rule executor and devs that way won't have to worry too much about the internals - they just know that you make 1 writer in the executor.

Maybe we could also add an optional parameter to `getWriter` to disable the rollover check, for use in route handlers that don't want to pay the overhead of the rollover check?

Reply:
> That's I'd say how it should be done. If you're in an executor function, you will get a `namespace` from the rule parameters and get the writer. Then you'll pass it down to your classes and functions as a dependency to write events.

Yeah, it *should* be done that way. But my point is that the current API design doesn't enforce that. And in fact, it is much easier to use the API in the wrong way. Currently, all classes/functions depend on `ruleDataClient` instead of `writer`, like [here](https://github.com/elastic/kibana/pull/102586/files#diff-141e4fdb5c577bce459bdfa51ef8d6265cef92b0e0132fb7664d20eff409f147R105), and [here](https://github.com/elastic/kibana/pull/102586/files#diff-24c750f51a9344fc040bb9b8e044a3920fb09b68ff69cce9b0cb226d5d1e1c15R240). And I'm also doing something similar in other places in my PR, just because it's much easier. I already have `ruleDataClient` as a dependency; why bothering extracting a `writer` and passing it down as another dependency? And some developers (including me 🙂) prefer to just copy-paste working code from other places instead of reading docs and exploring how things work under the hood. That's why I think adding docs wouldn't make much difference. That said, it's still a good practice to have docs.

I think a well-designed system should make it easy to do the right things and hard to do the wrong things. That's why I'm highlighting the current API as it potentially could be misused. But, as we have plans to improve RuleDataClient in a follow PR, let's discuss the API design separately. It is not a blocker for current changes to be merged.

Reply:
I will try to address this concern in the next PR within https://github.com/elastic/kibana/issues/101016 (will be a parallel PR to this one). The idea is to bind `RuleDataClient` to a namespace, so you could get an instance of it only if you know the namespace you're gonna work with:

```
// somewhere in the setup phase of security_solution
await ruleDataService.bootstrap(detectionAlertsLogDefinition);

// somewhere in an executor or a route handler
const detectionAlertsClient = await ruleDataService.getClient(detectionAlertsLogDefinition, namespace);

// or we could even make it sync
const detectionAlertsClient = ruleDataService.getClient(detectionAlertsLogDefinition, namespace);

const writer = detectionAlertsClient.getWriter();
const reader = detectionAlertsClient.getReader();
```

@marshallmain @xcrzx